src/RingArrays.jl

```diff
@@ -1,5 +1,236 @@
 module RingArrays
 
-# package code goes here
+using VirtualArrays
+using Iterators
 
-end # module
+import Base: size, getindex, checkbounds, display, RefValue, showerror
+export RingArray, size, checkbounds, display
+export getindex, load_block, RingArrayBoundsError
+export showerror, OverwriteError, RingArrayFullError, RingArrayBoundsError
+
+type RingArray{T, N} <: AbstractArray{T, N}
+    next_write::Int
+    max_blocks::Int
+    blocks::Array{AbstractArray{T, N}, 1}
+    num_users::Array{Int, 1}
+    block_size::Tuple
+    block_length::Int
+    range::UnitRange{Int}
+    data_length::Int
+
+    function RingArray(;max_blocks::Int=10, block_size::Tuple=(10,), data_length::Int=100)
+        if length(block_size) != N
+            throw(DimensionMismatch("block size $(block_size) does not match dimension of RingArray $(N)"))
+        end
+        return new(1, max_blocks,
+            Array{AbstractArray{T, N}, 1}(max_blocks),
+            zeros(Int, max_blocks), block_size,
+            block_size[1], 1:0, data_length)
+    end
+end
+
+type OverwriteError <: Exception
+    ring::RingArray
+end
+
+type RingArrayBoundsError <: Exception
+    ring::RingArray
+    i
+end
+
+type RingArrayFullError <: Exception
+    ring::RingArray
+end
+
+function showerror(io::IO, err::OverwriteError)
+    next_write = err.ring.next_write
+    num_users = err.ring.num_users[err.ring.next_write]
+    print(io, "OverwriteError: Cannot overwrite block $(next_write) since it has $(num_users) views")
+end
+
+function showerror(io::IO, err::RingArrayBoundsError)
+    range = tuple(err.ring.range, err.ring.block_size[2:end]...)
+    print(io, "RingArrayBoundsError: Cannot index $(err.i), outside of range $range")
+end
+
+function showerror(io::IO, err::RingArrayFullError)
+    print(io, "RingArrayFullError: Cannot load another block, max data length is $(err.ring.data_length)")
+end
+
+function display(ring::RingArray)
+    display(ring.blocks)
+end
+
+function size{T, N}(ring::RingArray{T, N})
+    if N <= 1
+        return tuple(ring.block_length * ring.max_blocks,)
+    else
+        return tuple(ring.block_length * ring.max_blocks, ring.block_size[2:end]...)
+    end
+end
+
+checkbounds(ring::RingArray) = true # because of warnings
```

Comment:
I imagine we could probably speed up these checkbounds functions a bit (@simd and @fastmath would probably be an easy start).